Add a cmdline option to add extra volume tags

jieyu opened the pull request:
**Is this a bug fix or adding new feature?**

New feature.

Fixes #333

**What is this PR about? / Why do we need it?**

Add a new cli command option `--extra-volume-tags` which is a map of
string to string (syntax is similar to `--node-labels` for kubelet).  By
default, it's an empty map which maps the the current behavior.

If this option is not empty, when doing dynamic provisioning (i.e., in
`CreateVolume`), we always attach the extra volume tags when calling
`d.cloud.CreateDisk`.

See why we need this in #333.

**What testing is done?** 

Yes. Unit tests have been added.

jieyu commented:
@leakingtapan PTAL

jieyu commented:
/assign @leakingtapan

jieyu commented:
Hum, looking at the coverage report. the coverage down is due to pkg/driver/node.go, which I haven't touched 🤔

leakingtapan commented:
Thx for the PR! Will spend some time review after 0.4 release is cut.

/hold

leakingtapan commented:
/ok-to-test

jieyu commented:
The previous integration test failure was due to https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/353/commits/4f947aafe9784aa1887a769dd835ce67234bd19a

jieyu commented:
/retest pull-aws-ebs-csi-driver-e2e-single-az

jieyu commented:
/retest

jieyu commented:
/test pull-aws-ebs-csi-driver-e2e-single-az

jieyu commented:
@leakingtapan a soft ping. Can you advise what's left to get this merged?

leakingtapan commented:
@jieyu Excuse my late reply. I was occupied with v0.4.0 release and some other things. I will review your change by the end of the week.

leakingtapan commented:
Also the coverage is dropped. Could you check if there is any new changed not covered?

leakingtapan commented:
/hold cancel

jieyu commented on pkg/cloud/cloud.go:
```diff
@@ -262,7 +274,9 @@ func (c *cloud) CreateDisk(ctx context.Context, volumeName string, diskOptions *
 	var tags []*ec2.Tag
 	for key, value := range diskOptions.Tags {
-		tags = append(tags, &ec2.Tag{Key: &key, Value: &value})
+		copiedKey := key
```
This is a bug in the old code. You cannot take a reference to the iteration variable. If you have multiple key value pairs (i.e., `diskOptions.Tags`), they end up pointing to the same object.

chrisglencross commented:
Hi - I'm interested in this change. Is any help required to complete it, for example addressing the small number of review comments?

jieyu commented on pkg/driver/node.go:
```diff
@@ -69,23 +69,25 @@ var (
+	metadata      cloud.MetadataService
+	mounter       Mounter
+	inFlight      *internal.InFlight
+	driverOptions *DriverOptions
```
Done.

jieyu commented on pkg/driver/validation_test.go:
```diff
@@ -0,0 +1,110 @@
+			expErr: fmt.Errorf("Volume tag key prefix '%s' is reserved", cloud.KubernetesTagKeyPrefix),
+		},
+		{
+			name: "invalid tag: reserved Kubernetes key prefix",
```
Done.

jieyu commented on pkg/driver/validation_test.go:
```diff
@@ -0,0 +1,110 @@
+			expErr: fmt.Errorf("Volume tag key too long (actual: %d, limit: %d)", cloud.MaxTagKeyLength+1, cloud.MaxTagKeyLength),
+		},
+		{
+			name: "invalid tag: key too long",
```
Thanks! done.

jieyu commented:
/test pull-aws-ebs-csi-driver-integration

jieyu commented:
@leakingtapan I addressed your comments and rebased the PR. The test failures look unrelated given other PR failed similarly.

I check the coverage of the new code. All new code is covered.

Sorry about the delay @chrisglencross

chrisglencross commented:
@jieyu No problem, and thank you!

leakingtapan commented:
@jieyu I fixed the test failures. Could you rebase from master? thx

jieyu commented:
@leakingtapan rebased. Please take a look!

leakingtapan commented:
/lgtm
/approve
